### PR TITLE
Block Patterns: Load basic core patterns from Pattern Directory

### PR DESCRIPTION
**Pull request by ryelle**
This ports over the pattern-directory API endpoint from Gutenberg trunk. This was added in https://github.com/WordPress/gutenberg/pull/28800 & https://github.com/WordPress/gutenberg/pull/26578.

I've removed most of the default patterns, and hooked up the Pattern Directory API to return the patterns marked "Core" from https://wordpress.org/patterns/.

This remote loading checks for `core-block-patterns` theme support and adds a new filter, `should_load_remote_block_patterns`. If `core-block-patterns` theme support is removed, it will disable all core patterns including remote patterns. If `should_load_remote_block_patterns` is filtered to return false, the core query blocks will be registered, but the remote request won't happen.

This is done without the `blockTypes` support, since https://github.com/WordPress/gutenberg/pull/32113 is still in progress. If I understand the process correctly, that can be done during feature-freeze?

Trac ticket: https://core.trac.wordpress.org/ticket/53246

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

File: src/wp-includes/rest-api/endpoints/class-wp-rest-pattern-directory-controller.php [hunk 1]
```diff
@@ -0,0 +1,297 @@
...
+		/**
+		 * Filters the REST API response for a pattern.
+		 *
+		 * @since 5.8.0
+		 *
+		 * @param WP_REST_Response $response    The response object.
+		 * @param object           $raw_pattern The unprepared pattern.
+		 * @param WP_REST_Request  $request     The request object.
+		 */
+		return apply_filters( 'rest_prepare_application_password', $response, $raw_pattern, $request );
```
Comment (iandunn) [inline src/wp-includes/rest-api/endpoints/class-wp-rest-pattern-directory-controller.php:184]: Doh, it looks like I missed this copy/pasta in https://github.com/WordPress/gutenberg/pull/26578/

**Comment by desrosj** [comment]
I started looking into the failing test here. It looks like if I comment out the added `current_screen` filter, the tests pass as expected. Also worth noting that the test passes when running only one specific group.

I have to step away for a bit, but I'll circle back when I return to continue looking into the issue.

**Comment by TimothyBJacobs** [comment]
These errors are occurring in `rest_preload_api_request` which is what is causing the preload data to be empty, and causing the test to fail.

```
WP_REST_Response Object
(
    [links:protected] => Array
        (
        )

    [matched_route:protected] => 
    [matched_handler:protected] => 
    [data] => Array
        (
            [code] => rest_no_route
            [message] => No route was found matching the URL and request method.
            [data] => Array
                (
                    [status] => 404
                )

        )

    [headers] => Array
        (
        )

    [status] => 404
)
WP_REST_Response Object
(
    [links:protected] => Array
        (
        )

    [matched_route:protected] => 
    [matched_handler:protected] => 
    [data] => Array
        (
            [code] => rest_no_route
            [message] => No route was found matching the URL and request method.
            [data] => Array
                (
                    [status] => 404
                )

        )

    [headers] => Array
        (
        )

    [status] => 404
)
```

Not sure why those routes don't appear to be registered yet. It may be that the REST Server initialization caused by the `_load_remote_block_patterns` is too early somehow?

**Comment by TimothyBJacobs** [comment]
I pushed a fix. Ideally we'd have some way of detecting when the REST API server needs to be registered built in to the base test class, but alas.

File: src/wp-includes/rest-api/endpoints/class-wp-rest-pattern-directory-controller.php (hunk 1 above)
Comment (ryelle) [reply src/wp-includes/rest-api/endpoints/class-wp-rest-pattern-directory-controller.php:184]: @iandunn should it be removed or just changed? `rest_prepare_block_pattern`?

Comment (iandunn) [reply src/wp-includes/rest-api/endpoints/class-wp-rest-pattern-directory-controller.php:184]: Probably changed, `rest_prepare_block_pattern` sounds good to me.

**Comment by ryelle** [comment]
Thanks for tracking that down, @TimothyBJacobs!

**Comment by ryelle** [comment]
Committed this first pass in [r51021](https://core.trac.wordpress.org/changeset/51021).

**Pull request closed without merging**